### PR TITLE
Use exact numDocs in synced-flush and metadata snapshot

### PR DESCRIPTION
**Pull request by dnhatn**
Since #29458, we use a searcher to calculate the number of documents for a commit stats. Sadly, that approach is flawed. The searcher might no longer point to the last commit if it's refreshed. As synced-flush requires an exact numDocs to work correctly, we have to exclude all soft-deleted docs.

This commit makes synced-flush stop using CommitStats but read an exact numDocs directly from an index commit.

Relates #29458
Relates #29530

### CODE REVIEW

**Comment by dnhatn** [comment]
@elasticmachine test this please

**Comment by bleskes** [comment]
I wonder if we should do this. Classically number of docs was always total docs in the segments. I'm not saying this isn't debatable but I'm also not sure it merits the breakage price. Also we want this to go into 6.x. Instead of this, maybe we should another field to the stats to reflect soft deletes? then sycned flush is freed to do what it thinks is right and we have more information.

**Review by s1monw — CHANGES_REQUESTED** [review]
left one comment

**Comment by dnhatn** [comment]
> I think we need a better utility method like Lucene. getNumDocs that takes soft deletes into account? 

@bleskes Yes, we should do it.

**Comment by dnhatn** [comment]
@s1monw I take a different approach for this. Can you please have a look? Thank you.

**Comment by s1monw** [comment]
I am ok with the changing the semantics here. I think we try to make something work that isn't necessarily giving us enough information. I think we should keep it the way it is and deprecate it in 6.x? I really wonder what the purpose of this stats is compared to other stats that have a live numDocs that is accurate?

**Comment by bleskes** [comment]
@s1monw as said, I think you are a better judge here for what's handy. I'm not sure I fully follow what you suggest.

>  I think we should keep it the way it is and deprecate it in 6.x? 

Double checking - the way it is in master or in the PR? can you clarify?

> other stats that have a live numDocs that is accurate?

What other stats are you referring to?

**Comment by s1monw** [comment]
> Double checking - the way it is in master or in the PR? can you clarify?

the way it is in master

> other stats that have a live numDocs that is accurate?

I mean for instance docStats here: https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-stats.html

**Comment by dnhatn** [comment]
@s1monw  and @bleskes,

I've updated the PR to restore the CommitStats and made synced-flush using an exact numDocs only when soft-deletes is enabled. There are still some issues that I am not sure.

1. With this change, only the synced-flush call uses an exact numDocs. Should we still need to cache this value in an engine?

2. In peer-recovery, a replica shard needs to send an exact numDocs to the primary shard if soft-deletes is enabled and an index was sealed. I see two possible approaches:

- Pass an extra parameter "requiredExactNumDocs" in `getStoreMetadataSnapshot`. Internally we always calculate an exact numDocs if this parameter is true. This does not optimize the case when the index was not sealed and may cause confusion for other StoreMetadataSnapshot's usages.

- Override the numDocs by an exact numDocs when the soft-deletes is true and an index was sealed. This is not nice but the change is minimal and does not affect StoreMetadataSnapshot's other usages.

Please have a look and let me know your thought. Thank you!

File: server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java [hunk 1]
```diff
@@ -289,9 +293,21 @@ public RecoveryResponse newInstance() {
...
      */
-    private Store.MetadataSnapshot getStoreMetadataSnapshot(final RecoveryTarget recoveryTarget) {
+    static Store.MetadataSnapshot getStoreMetadataSnapshot(final Logger logger, final RecoveryTarget recoveryTarget) {
         try {
-            return recoveryTarget.indexShard().snapshotStoreMetadata();
+            final Store.MetadataSnapshot snapshot = recoveryTarget.indexShard().snapshotStoreMetadata();
+            // The primary shard may need the "exact" numDocs to verify if the commit has syncId.
+            final boolean softDeleteEnabled = recoveryTarget.indexShard().indexSettings().isSoftDeleteEnabled();
+            if (softDeleteEnabled && Strings.hasText(snapshot.getSyncId())) {
+                final List<IndexCommit> commits = DirectoryReader.listCommits(recoveryTarget.store().directory());
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java]: This needs to use safe commit. This is the commit the engine will use (probably an existing bug) .

**Comment by bleskes** [comment]
Thx @dnhatn . I like where this is going. IMO neither synced flush nor recover are frequent operations. We should just use exact numbers all the time and everywhere. keep it simple. if soft deletes are not enabled, it's the same as the "fast" numDocs (and we can assert for that).

File: server/src/main/java/org/elasticsearch/common/lucene/Lucene.java [hunk 2]
```diff
@@ -149,6 +150,16 @@ public static int getNumDocs(SegmentInfos info) {
         return numDocs;
     }
 
+    /**
+     * Unlike {@link #getNumDocs(SegmentInfos)} this method returns a numDocs that always excludes soft-deleted docs.
+     * This method is expensive thus prefer using {@link #getNumDocs(SegmentInfos)} unless an exact numDocs is required.
+     */
+    public static int getExactNumDocs(IndexCommit commit) throws IOException {
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/common/lucene/Lucene.java]: ++

File: server/src/main/java/org/elasticsearch/indices/flush/SyncedFlushService.java [hunk 3]
```diff
@@ -466,15 +468,24 @@ public String executor() {
...
         indexShard.flush(flushRequest);
-        final CommitStats commitStats = indexShard.commitStats();
-        final Engine.CommitId commitId = commitStats.getRawCommitId();
-        logger.trace("{} pre sync flush done. commit id {}, num docs {}", request.shardId(), commitId, commitStats.getNumDocs());
-        return new PreSyncedFlushResponse(commitId, commitStats.getNumDocs(), commitStats.syncId());
+        try (Engine.IndexCommitRef commitRef = indexShard.acquireLastIndexCommit(false)) {
+            final SegmentInfos segmentInfos = Lucene.readSegmentInfos(commitRef.getIndexCommit());
+            final int numDocs;
+            if (indexShard.indexSettings().isSoftDeleteEnabled()) {
+                numDocs = Lucene.getExactNumDocs(commitRef.getIndexCommit());
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/indices/flush/SyncedFlushService.java]: you can call the `getExactNumDocs` all the time. it has no overhead if soft_deletes are not there

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 4]
```diff
@@ -341,7 +342,8 @@ public void phase1(final IndexCommit snapshot, final Supplier<Integer> translogO
                     recoverySourceSyncId.equals(recoveryTargetSyncId);
             if (recoverWithSyncId) {
                 final long numDocsTarget = request.metadataSnapshot().getNumDocs();
-                final long numDocsSource = recoverySourceMetadata.getNumDocs();
+                final boolean softDeletesEnabled = shard.indexSettings().isSoftDeleteEnabled();
+                final long numDocsSource = softDeletesEnabled ? Lucene.getExactNumDocs(snapshot) : recoverySourceMetadata.getNumDocs();
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: same here just use `getExactNumDocs`

**Review by s1monw — APPROVED** [review]
LGTM

**Comment by dnhatn** [comment]
@elasticmachine test this please

**Review by bleskes — APPROVED** [review]
LGTM. Thanks Nhat.

**Comment by dnhatn** [comment]
Thanks @bleskes and @s1monw

**Pull request merged**